session_rpcserver.go

```diff
@@ -20,33 +21,39 @@ type sessionRpcServer struct {
 
 	db            *session.DB
 	sessionServer *session.Server
+
+	superMacBaker func(ctx context.Context, rootKeyID uint64,
+		recipe *session.MacaroonRecipe) (string, error)
 }
 
 // AddSession adds and starts a new Terminal Connect session.
 func (s *sessionRpcServer) AddSession(_ context.Context,
 	req *litrpc.AddSessionRequest) (*litrpc.AddSessionResponse, error) {
 
-	var (
-		typ    session.Type
-		expiry time.Time
-	)
-	switch req.SessionType {
-	case litrpc.SessionType_TYPE_UI_PASSWORD:
-		typ = session.TypeUIPassword
+	typ, err := unmarshalRPCType(req.SessionType)
+	if err != nil {
+		return nil, err
+	}
 
-	default:
+	if typ != session.TypeUIPassword && typ != session.TypeMacaroonAdmin {
 		return nil, fmt.Errorf("invalid session type, only UI " +
-			"password supported in LiT")
+			"password, macaroon admin and macaroon readonly " +
+			"types supported in LiT")
 	}
 
-	expiry = time.Unix(int64(req.ExpiryTimestampSeconds), 0)
+	expiry := time.Unix(int64(req.ExpiryTimestampSeconds), 0)
 	if time.Now().After(expiry) {
 		return nil, fmt.Errorf("expiry must be in the future")
 	}
 
+	var macPerms []bakery.Op
+	if typ == session.TypeMacaroonAdmin {
+		macPerms = getAllPermissions()
```

Comment:
I'm still trying to decide if we should define these permissions at _session create time_ or at _bake time_. E.g. if we add a new permission to one of the daemons, does the user expect the session to upgrade automatically (since it's an admin or full read-only session) after restart? Or would the user need to create a new session if something in the permissions changed? The first would be better UX while the second might be a bit more security conscious.

The difference in the code would be: for `TypeMacaroonAdmin` and `TypeMacaroonReadonly` we wouldn't store the recipe in the DB (that would be reserved for the future `TypeMacaroonCustom`) and assemble the list of permissions right before baking the macaroon.

What do you think?

Reply:
hmmmm yeah I agree with your pro's and con's list of the two options. I guess option 1 is ok given that if the user wants to be more security focused and not let the perms list auto-update, they should use the custom option

Reply:
option 1 also makes the db migration easier since all we then need to do is change the type & dont need to go populate the recipe field